### PR TITLE
*: shard implicit row ID

### PR DESCRIPTION
**Pull request by coocood**
Shard the implicit row ID to avoid hot spot.

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -828,8 +828,22 @@ func handleTableOptions(options []*ast.TableOption, tbInfo *model.TableInfo) {
 			tbInfo.Charset = op.StrValue
 		case ast.TableOptionCollate:
 			tbInfo.Collate = op.StrValue
+		case ast.TableOptionShardRowID:
+			if !hasAutoIncrementColumn(tbInfo) {
+				//tbInfo.ShardRowID = op.UintValue != 0
```
Comment (winkyao) [inline ddl/ddl_api.go]: remove this?

File: executor/show.go [hunk 2]
```diff
@@ -623,6 +623,10 @@ func (e *ShowExec) fetchShowCreateTable() error {
 		}
 	}
 
+	if tb.Meta().ShardRowID {
+		buf.WriteString(" SHARD_ROW_ID=1")
```
Comment (winkyao) [inline executor/show.go]: if we enable shard_row_id by `SHARD_ROW_ID=2`?

File: parser/parser.y [hunk 3]
```diff
@@ -5272,6 +5273,10 @@ TableOption:
 	{
 		$$ = &ast.TableOption{Tp: ast.TableOptionStatsPersistent}
 	}
+|	"SHARD_ROW_ID" EqOpt LengthNum
```
Comment (winkyao) [inline parser/parser.y]: is `EqOpt LengthNum` necessary?

File: executor/show.go (hunk 2 above)
Comment (coocood) [reply executor/show.go]: It's enabled too.

File: parser/parser.y (hunk 3 above)
Comment (coocood) [reply parser/parser.y]: So we can disable it after enabled.

**Comment by coocood** [comment]
@ngaut @shenli PTAL

**Comment by coocood** [comment]
@zhangjinpeng1987 PTAL

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -932,6 +956,36 @@ func (d *ddl) RebaseAutoID(ctx context.Context, ident ast.Ident, newBase int64)
 	return errors.Trace(err)
 }
 
+func (d *ddl) ShardRowID(ctx context.Context, ident ast.Ident, uVal uint64) error {
```
Comment (shenli) [inline ddl/ddl_api.go]: add comment for this function

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -932,6 +956,36 @@ func (d *ddl) RebaseAutoID(ctx context.Context, ident ast.Ident, newBase int64)
...
+		return errors.Trace(err)
+	}
+	job.Type = model.ActionShardRowID
+	job.Args = []interface{}{uVal}
+	err = d.doDDLJob(ctx, job)
+	err = d.callHookOnChanged(err)
+	return errors.Trace(err)
+}
+
+func (d *ddl) createJobByIdent(ident ast.Ident) (*model.Job, error) {
```
Comment (shenli) [inline ddl/ddl_api.go]: createJobByName is a better name?

File: table/tables/tables.go [hunk 6]
```diff
@@ -698,7 +700,26 @@ func GetColDefaultValue(ctx context.Context, col *table.Column, defaultVals []ty
...
+		rowID |= *txnCtx.Shard
+	}
+	return rowID, nil
+}
+
+func (t *Table) calcShard(startTS uint64) int64 {
+	var buf [8]byte
+	binary.LittleEndian.PutUint64(buf[:], startTS)
+	hashVal := int64(murmur3.Sum32(buf[:]))
+	return (hashVal & (1<<t.meta.ShardRowIDBits - 1)) << (64 - t.meta.ShardRowIDBits - 1)
```
Comment (shenli) [inline table/tables/tables.go]: Could we make the `(1<<t.meta.ShardRowIDBits - 1)) << (64 - t.meta.ShardRowIDBits - 1)` a const?

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -932,6 +956,36 @@ func (d *ddl) RebaseAutoID(ctx context.Context, ident ast.Ident, newBase int64)
...
+	is := d.GetInformationSchema()
+	schema, ok := is.SchemaByName(ident.Schema)
+	if !ok {
+		return nil, infoschema.ErrDatabaseNotExists.GenByArgs(ident.Schema)
+	}
+	t, err := is.TableByName(ident.Schema, ident.Name)
+	if err != nil {
+		return nil, errors.Trace(infoschema.ErrTableNotExists.GenByArgs(ident.Schema, ident.Name))
```
Comment (zimulala) [inline ddl/ddl_api.go]: Can we remove this trace ？

File: table/tables/tables.go (hunk 6 above)
Comment (coocood) [reply table/tables/tables.go]: No, `t.meta.ShardRowIDBits` is different for each table.

File: ddl/ddl_api.go (hunk 5 above)
Comment (coocood) [reply ddl/ddl_api.go]: `ident` has both `Schema` and `TableName`, it's not a single name.

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -888,9 +904,17 @@ func (d *ddl) AlterTable(ctx context.Context, ident ast.Ident, specs []*ast.Alte
 			err = ErrUnsupportedModifyPrimaryKey.GenByArgs("drop")
 		case ast.AlterTableOption:
 			for _, opt := range spec.Options {
-				if opt.Tp == ast.TableOptionAutoIncrement {
+				switch opt.Tp {
+				case ast.TableOptionShardRowID:
+					if opt.UintValue > 15 {
+						opt.UintValue = 15
```
Comment (zimulala) [inline ddl/ddl_api.go]: Can we define a constant as 15? We use it in multiple places.

**Comment by coocood** [comment]
@zimulala @shenli PTAL

File: table/tables/tables.go [hunk 9]
```diff
@@ -698,7 +700,26 @@ func GetColDefaultValue(ctx context.Context, col *table.Column, defaultVals []ty
...
+		}
+		rowID |= *txnCtx.Shard
+	}
+	return rowID, nil
+}
+
+func (t *Table) calcShard(startTS uint64) int64 {
+	var buf [8]byte
+	binary.LittleEndian.PutUint64(buf[:], startTS)
+	hashVal := int64(murmur3.Sum32(buf[:]))
```
Comment (zimulala) [inline table/tables/tables.go]: Why not use `buf` directly?

File: ddl/ddl_api.go (hunk 5 above)
Comment (shenli) [reply ddl/ddl_api.go]: createJobForTable

File: executor/ddl_test.go [hunk 10]
```diff
@@ -334,3 +335,29 @@ func (s *testSuite) TestTooLargeIdentifierLength(c *C) {
...
+	c.Assert(err, IsNil)
+	var hasShardedID bool
+	var count int
+	c.Assert(tk.Se.NewTxn(), IsNil)
+	err = tbl.IterRecords(tk.Se, tbl.FirstKey(), nil, func(h int64, rec []types.Datum, cols []*table.Column) (more bool, err error) {
+		c.Assert(h, GreaterEqual, int64(0))
+		if (h >> 56) > 0 {
+			hasShardedID = true
+		}
+		count++
```
Comment (zimulala) [inline executor/ddl_test.go]: Can we put this line to line 355？ Then we can make sure all IDs are `sharded ID`.

File: executor/ddl_test.go [hunk 11]
```diff
@@ -334,3 +335,29 @@ func (s *testSuite) TestTooLargeIdentifierLength(c *C) {
...
+		tk.MustExec(fmt.Sprintf("insert t values (%d)", i))
+	}
+	tbl, err := domain.GetDomain(tk.Se).InfoSchema().TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	var hasShardedID bool
+	var count int
+	c.Assert(tk.Se.NewTxn(), IsNil)
+	err = tbl.IterRecords(tk.Se, tbl.FirstKey(), nil, func(h int64, rec []types.Datum, cols []*table.Column) (more bool, err error) {
+		c.Assert(h, GreaterEqual, int64(0))
+		if (h >> 56) > 0 {
```
Comment (zimulala) [inline executor/ddl_test.go]: Why there is `56`? Can we use a temporary variable to replace it? Then make it more readable.

File: executor/ddl_test.go (hunk 10 above)
Comment (coocood) [reply executor/ddl_test.go]: No, sharded ID still include shard 0, then `(h >> 56) == 0`.

**Comment by coocood** [comment]
@shenli @zimulala PTAL

File: ddl/ddl_api.go (hunk 5 above)
Comment (shenli) [reply ddl/ddl_api.go]: createJobForTable is a better name.

File: meta/autoid/autoid.go [hunk 12]
```diff
@@ -27,7 +27,7 @@ import (
 )
 
 // Test needs to change it, so it's a variable.
-var step = int64(5000)
+var step = int64(30000)
```
Comment (shenli) [inline meta/autoid/autoid.go]: We should add a metrics for auto-id allocation latter.

File: executor/show.go [hunk 13]
```diff
@@ -623,6 +623,10 @@ func (e *ShowExec) fetchShowCreateTable() error {
 		}
 	}
 
+	if tb.Meta().ShardRowIDBits > 0 {
+		buf.WriteString(fmt.Sprintf(" SHARD_ROW_ID_BITS=%d", tb.Meta().ShardRowIDBits))
```
Comment (shenli) [inline executor/show.go]: We should also use special-comment for `show create table`. Or users will meet error when migrating data from TiDB to MySQL.

File: sessionctx/binloginfo/binloginfo.go [hunk 14]
```diff
@@ -108,3 +111,19 @@ func SetDDLBinlog(client interface{}, txn kv.Transaction, jobID int64, ddlQuery
...
+
+const specialPrefix = `/*!90000 `
+
+func addSpecialComment(ddlQuery string) string {
+	if strings.Contains(ddlQuery, specialPrefix) {
+		return ddlQuery
+	}
+	upperQuery := strings.ToUpper(ddlQuery)
+	reg, err := regexp.Compile(`SHARD_ROW_ID_BITS\s*=\s*\d+`)
+	terror.Log(err)
```
Comment (zimulala) [inline sessionctx/binloginfo/binloginfo.go]: If err isn't nil, reg may be nil.

Comment (coocood) [reply sessionctx/binloginfo/binloginfo.go]: It's just added to pass `errcheck`, it never returns nil

**Comment by zimulala** [comment]
LGTM

**Comment by shenli** [comment]
LGTM
Please resolve the conflicts.

**Pull request merged**